src/main/java/org/nuxeo/audit/storage/impl/SQLAuditStorage.java

```diff
@@ -0,0 +1,196 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Contributors:
+ *     Estelle Giuly <egiuly@nuxeo.com>
+ */
+
+package org.nuxeo.audit.storage.impl;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.codehaus.jackson.annotate.JsonIgnore;
+import org.codehaus.jackson.annotate.JsonProperty;
+import org.codehaus.jackson.map.ObjectMapper;
+import org.nuxeo.ecm.core.api.CursorService;
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.api.DocumentModelList;
+import org.nuxeo.ecm.core.api.DocumentRef;
+import org.nuxeo.ecm.core.api.ScrollResult;
+import org.nuxeo.ecm.core.query.sql.model.MultiExpression;
+import org.nuxeo.ecm.core.query.sql.model.OrderByExpr;
+import org.nuxeo.ecm.core.query.sql.model.Predicate;
+import org.nuxeo.ecm.core.query.sql.model.StringLiteral;
+import org.nuxeo.ecm.directory.Session;
+import org.nuxeo.ecm.directory.api.DirectoryService;
+import org.nuxeo.ecm.directory.sql.SQLDirectory;
+import org.nuxeo.ecm.directory.sql.SQLSession;
+import org.nuxeo.ecm.platform.audit.api.AuditQueryBuilder;
+import org.nuxeo.ecm.platform.audit.api.AuditStorage;
+import org.nuxeo.ecm.platform.audit.api.ExtendedInfo;
+import org.nuxeo.ecm.platform.audit.api.LogEntry;
+import org.nuxeo.ecm.platform.audit.impl.LogEntryImpl;
+import org.nuxeo.runtime.api.Framework;
+
+/**
+ * Audit storage implementation for a SQL Audit directory.
+ * 
+ * @since 9.3
+ */
+public class SQLAuditStorage implements AuditStorage {
+
+	public static final String NAME = "SQLAuditStorage";
+
+	protected static final Log log = LogFactory.getLog(SQLAuditStorage.class);
+
+	protected static final String DIRECTORY_NAME = "sqlAuditDirectory";
+
+	protected static final String ID_COLUMN = "id";
+
+	protected static final String JSON_COLUMN = "entry";
+
+	protected CursorService<Iterator<LogEntry>, LogEntry> cursorService = new CursorService<>();
+
+	protected SQLDirectory sqlDirectory = (SQLDirectory) Framework.getService(DirectoryService.class)
+			.getDirectory(DIRECTORY_NAME);
+
+	/**
+	 * Insert entries as Json in the SQL Audit directory.
+	 */
+	@Override
+	public void append(List<String> jsonEntries) {
+		Session session = sqlDirectory.getSession();
+		for (String jsonEntry : jsonEntries) {
+			Map<String, Object> jsonMap = new HashMap<String, Object>();
+			jsonMap.put(JSON_COLUMN, jsonEntry);
+			session.createEntry(jsonMap);
```

Comment:
You can probably do `session.createEntry(Collections.singletonMap(JSON_COLUMN, jsonEntry))` to simplify.

And actually rewrite the whole `for` loop as just:

    jsonEntries.forEach(jsonEntry -> session.createEntry(Collections.singletonMap(JSON_COLUMN, jsonEntry)));

Reply:
actually it creates a
```
java.lang.UnsupportedOperationException
	at java.util.AbstractMap.put(AbstractMap.java:209)
	at org.nuxeo.ecm.directory.sql.SQLSession.createEntryWithoutReferences(SQLSession.java:821)
```

Reply:
Ah ok. That's a bug in `SQLSession`, we should really make defensive copies if we modify passed-in structures. But that's for another ticket.

Reply:
FYI https://jira.nuxeo.com/browse/NXP-23722

Reply:
ah cool thanks